// @eduzz/rabbit/dist/test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Connection_1 = require("./Connection");
var fn_1 = require("./fn");
var Memory_1 = require("./Fallback/Adapter/Memory");
var connection = new Connection_1.Connection({
    dsn: 'amqp://guest:guest@localhost:5672/',
    exchange: 'test',
    exchangeType: 'topic',
    connectionName: 'test',
    maxConnectionAttempts: 10,
    processExitWhenUnableToConnectFirstTime: false,
});
connection.setFallbackAdapter(new Memory_1.Memory());
connection
    .queue('listener.fallback')
    .topic('listen.topic')
    .durable(true)
    .prefetch(10)
    .retryTimeout(60000)
    .listen(function (message) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                console.log('received:', message);
                return [4 /*yield*/, fn_1.sleep(4000)];
            case 1:
                _a.sent();
                return [2 /*return*/, true];
        }
    });
}); });
connection.queue('queue.without.consumer').topic('listen.topic').durable(true).create();
connection.delayQueue('asasdd.asdasdasd').durable().from('ouvindo.xpto').to('depois.do.delay').timeout(5000).create();
(function () { return __awaiter(void 0, void 0, void 0, function () {
    var publisher, anotherPublisher, result, anotherResult;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                publisher = connection.topic('listen.topic').persistent();
                anotherPublisher = connection.topic('another.topic').persistent();
                _a.label = 1;
            case 1:
                if (!true) return [3 /*break*/, 5];
                return [4 /*yield*/, publisher.send({ payload: 'message', priority: 1 })];
            case 2:
                result = _a.sent();
                return [4 /*yield*/, anotherPublisher.send({ payload: 'another.message', priority: 1 })];
            case 3:
                anotherResult = _a.sent();
                console.log({ result: result, anotherResult: anotherResult });
                return [4 /*yield*/, fn_1.sleep(5000)];
            case 4:
                _a.sent();
                return [3 /*break*/, 1];
            case 5: return [2 /*return*/];
        }
    });
}); })();
console.log('started');
