// mercurius-chat/dist/socket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disconnect = exports.send = exports.typing = exports.join = exports.connection = exports.persistData = void 0;
const tslib_1 = require("tslib");
const sentry = (0, tslib_1.__importStar)(require("@sentry/node"));
const rabbitmq_1 = require("./rabbitmq");
const persistData = async (params) => {
    try {
        await (0, rabbitmq_1.sendMessageToRabbit)(params);
    }
    catch (error) {
        sentry.captureException(error);
    }
};
exports.persistData = persistData;
const connection = (io, listener) => {
    console.log('Connection called');
    return io.on('connection', socket => {
        socket.on('connect_error', err => {
            console.log(`connect_error due to ${err.message}`);
        });
        console.log('CONNECTION');
        listener(socket);
    });
};
exports.connection = connection;
const join = (io, callback) => (0, exports.connection)(io, socket => {
    console.log('listener do join');
    socket.on('join', (payload) => {
        console.log('event do join');
        socket.join(String(payload.room));
        callback && callback(payload);
    });
});
exports.join = join;
const typing = (io, callback) => (0, exports.connection)(io, socket => {
    console.log('listener do typing');
    socket.on('typing', (payload) => {
        console.log('event do typing');
        socket.to(String(payload.room)).emit('typing', payload);
        callback && callback(payload);
    });
});
exports.typing = typing;
const send = (io, callback) => (0, exports.connection)(io, socket => {
    console.log('listener do send');
    socket.on('send', (payload) => {
        console.log('event do send: ', JSON.stringify(payload));
        socket.to(String(payload.room)).emit('receive', payload);
        if (!!payload.persistData) {
            (0, exports.persistData)(Object.assign(Object.assign({}, payload.persistData), { payload: payload.data }));
        }
        callback && callback(payload);
    });
});
exports.send = send;
const disconnect = (io, callback) => (0, exports.connection)(io, socket => {
    socket.on('disconnect', (payload) => {
        socket.to(String(payload.room)).emit('disconnect', payload);
        callback && callback(payload);
    });
});
exports.disconnect = disconnect;
exports.default = { connection: exports.connection, join: exports.join, send: exports.send, typing: exports.typing, disconnect: exports.disconnect };
